packages/shared/src/components/fields/MarkdownInput/CommentMarkdownInput.tsx

```diff
@@ -0,0 +1,209 @@
+import React, {
+  FormEventHandler,
+  ReactElement,
+  useEffect,
+  useMemo,
+  useRef,
+} from 'react';
+import { useMutation, useQueryClient } from 'react-query';
+import classNames from 'classnames';
+import { defaultMarkdownCommands } from '../../../hooks/input';
+import MarkdownInput, { MarkdownRef } from './index';
+import {
+  Comment,
+  COMMENT_ON_COMMENT_MUTATION,
+  COMMENT_ON_POST_MUTATION,
+  EDIT_COMMENT_MUTATION,
+  PostCommentsData,
+} from '../../../graphql/comments';
+import { graphqlUrl } from '../../../lib/config';
+import { useRequestProtocol } from '../../../hooks/useRequestProtocol';
+import { formToJson } from '../../../lib/form';
+import { getPostByIdKey, updatePostCache } from '../../../hooks/usePostById';
+import { PostData } from '../../../graphql/posts';
+import { useBackgroundRequest } from '../../../hooks/companion';
+import { Edge } from '../../../graphql/common';
+
+export interface CommentClassName {
+  container?: string;
+  tab?: string;
+}
+
+export interface CommentMarkdownInputProps {
+  postId?: string;
+  sourceId?: string;
+  editCommentId?: string;
+  parentCommentId?: string;
+  initialContent?: string;
+  replyTo?: string;
+  className?: CommentClassName;
+  onCommented?: (comment: Comment, isNew?: boolean) => void;
+}
+
+interface SubmitComment {
+  id?: string;
+  content: string;
+}
+
+const generateCommentEdge = (comment: Comment): Edge<Comment> => ({
+  node: { ...comment, children: { edges: [], pageInfo: null } },
+});
+
+export function CommentMarkdownInput({
+  postId,
+  sourceId,
+  parentCommentId,
+  initialContent,
+  replyTo,
+  editCommentId,
+  className = {},
+  onCommented,
+}: CommentMarkdownInputProps): ReactElement {
+  const client = useQueryClient();
+  const markdownRef = useRef<MarkdownRef>();
+  const key = useMemo(
```

Comment:
This seems pretty intense, should we not consider this a global key function?
 (which we have for some other parts)

Reply:
Right. Let's use that one 🚢